eegnb/experiments/visual_n170/n170.py

```diff
@@ -17,107 +17,40 @@
-    stim = [houses, faces]
-
-    # Show the instructions screen
-    show_instructions(duration)
-
-    if eeg:
-        if save_fn is None:  # If no save_fn passed, generate a new unnamed save file
-            random_id = random.randint(1000,10000)
-            save_fn = generate_save_fn(eeg.device_name, "visual_n170", random_id, random_id, "unnamed")
-            print(
-                f"No path for a save file was passed to the experiment. Saving data to {save_fn}"
-            )
-        eeg.start(save_fn, duration=record_duration + 5)
-
-    # Start EEG Stream, wait for signal to settle, and then pull timestamp for start point
-    start = time()
-
-    # Iterate through the events
-    for ii, trial in trials.iterrows():
-        # Inter trial interval
-        core.wait(iti + np.random.rand() * jitter)
-
-        # Select and display image
-        label = trials["image_type"].iloc[ii]
-        image = choice(faces if label == 1 else houses)
-        image.draw()
-
-        # Push sample
-        if eeg:
-            timestamp = time()
-            if eeg.backend == "muselsl":
-                marker = [markernames[label]]
-            else:
-                marker = markernames[label]
-            eeg.push_sample(marker=marker, timestamp=timestamp)
 
-        mywin.flip()
-
-        # offset
-        core.wait(soa)
-        mywin.flip()
-        if len(event.getKeys()) > 0 or (time() - start) > record_duration:
-            break
-
-        event.clearEvents()
-
-    # Cleanup
-    if eeg:
-        eeg.stop()
-
-    mywin.close()
-
-
-def show_instructions(duration):
-
-    instruction_text = """
-    Welcome to the N170 experiment! 
- 
-    Stay still, focus on the centre of the screen, and try not to blink. 
-
-    This block will run for %s seconds.
-
-    Press spacebar to continue. 
+    return [houses, faces]
     
-    """
-    instruction_text = instruction_text % duration
+def present_stimulus(trials, ii, eeg, markernames):
+   
+    label = trials["image_type"].iloc[ii]
+    image = choice(faces if label == 1 else houses)
+    image.draw()
 
-    # graphics
-    mywin = visual.Window([1600, 900], monitor="testMonitor", units="deg", fullscr=True)
-
-    mywin.mouseVisible = False
-
-    # Instructions
-    text = visual.TextStim(win=mywin, text=instruction_text, color=[-1, -1, -1])
-    text.draw()
-    mywin.flip()
-    event.waitKeys(keyList="space")
+    # Push sample
+    if eeg:
+        timestamp = time()
+        if eeg.backend == "muselsl":
+            marker = [markernames[label]]
+        else:
+            marker = markernames[label]
+        eeg.push_sample(marker=marker, timestamp=timestamp)
+   
+
+if __name__ == "__main__":
+     
+    test = Experiment("Visual N170")
+    test.instruction_text = instruction_text
+    test.load_stimulus = load_stimulus
+    test.present_stimulus = present_stimulus
+    test.setup()
+    test.present()
```

Comment:
Why do it like this? I assumed you'd create a subclass of Experiment?

Reply:
I think this works better because I don't need to care about redefining a bunch of stuff and thinking about defining the functions in the base class etc. This way I can use the same class everywhere and just initialize it differently for the different experiments. And since I am seperating the present_stimulus and load_stimulus functions, it does not make much of a difference

Reply:
I agree with Erik I think the natural way of doing this is with an experiment base class that is subclassed for every experiment. The functions that you are passing here would then be (potentially decorated) methods in the subclasses. 

You're right that you're saving some overhead by using the same class everywhere and not needing placeholder function definitions in the base class like 

```python
def present_stimulus(self):
  pass
```

...but that's pretty standard idiomatic python and isn't a huge deal. 


I'm still going through this thoroughly but I'd like a bit more spelling out as to why this function passing approach is a better than subclassing from a general code base structure perspective.